swf/src/read.rs

```diff
@@ -2123,7 +2123,31 @@ impl<'a> Reader<'a> {
                 ratio: self.read_u8()?,
             });
         }
-        Ok(GradientFilter {
+        Ok(GradientBevelFilter {
+            colors: gradient_records,
+            blur_x: self.read_fixed16()?,
+            blur_y: self.read_fixed16()?,
+            angle: self.read_fixed16()?,
+            distance: self.read_fixed16()?,
+            strength: self.read_fixed8()?,
+            flags: GradientFilterFlags::from_bits_truncate(self.read_u8()?),
+        })
+    }
+
+    fn read_gradient_glow_filter(&mut self) -> Result<GradientGlowFilter> {
```

Comment:
`read_gradient_bevel_filter` and `read_gradient_glow_filter` are exactly the same. Why both are needed?

Reply:
I had to split up GradientFilter into two different types to allow for different impls on the types (for example, GradientGlowFilter.to_avm2() gives a different thing to GradientBevelFilter.to_avm2())

Reply:
I see `FilterAvm2Ext` for the `Filter` enum + each `swf` filter type. So maybe we can just implement `from_avm2_object` and `as_avm2_object` for the `Filter` enum, without `FilterAvm2Ext` at all? I prefer to avoid this code duplication...

Reply:
Actually this would be unnecessary if we implemented the filters as native objects with getter/setter for each field. I start to feel like the object system in avm2 needs a refactor, since currently it's not so easy to add new native objects.

Reply:
That would be... very complicated.

These types do not exist in avm2 land, they are not exclusive to avm2, and they're defined in the renderer which has no concept of avmness at all. An `Ext` trait is required to put anything on them at all that interacts with Avm2.

You can have custom Object types for them, of course, but it'd need to store a **clone** of the filter object and nothing more. The objects avm2 gets are snapshots at that moment, not the actual direct filter on the stage.

Reply:
Okay, so we can keep `FilterAvm2Ext` but implement it just for `Filter`, leaving the conversion of each specific filter type to a free function - `bevel_filter_to_avm2`, `glow_filter_to_avm2` etc.

Regarding the native objects, we can store just a reference to the underlying `DisplayObject`, then make the getters/setters operate directly on the `DisplayObject`, so any filter updates will be reflected immediately. Alternatively, we can use interior mutability to allow both the `DisplayObject` and avm2 `Object` "own" the filters. Ideally I think native objects are the way to go, as this pattern is used very widely in avmplus.